src/librustc_resolve/lib.rs

```diff
@@ -2128,7 +2131,11 @@ impl<'a> Resolver<'a> {
         // An invalid forward use of a type parameter from a previous default.
         if let ForwardTyParamBanRibKind = all_ribs[rib_index].kind {
             if record_used {
-                self.report_error(span, ResolutionError::ForwardDeclaredTyParam);
+                if rib_ident.name == kw::SelfUpper {
```

Comment:
```suggestion
                let res_error = match rib_ident.name {
```
(I'd move out the call to `self.report_error`)